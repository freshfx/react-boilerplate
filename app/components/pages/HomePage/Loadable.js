/**
 * Asynchronously loads the component for HomePage
 */
import React from 'react'

import LoadingIndicator from 'components/atoms/LoadingIndicator'
import withLoadable from 'utils/withLoadable'

export default withLoadable(() => import('./index'), {
  fallback: <LoadingIndicator />
})
