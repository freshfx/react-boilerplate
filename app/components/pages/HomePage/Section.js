import styled from '@emotion/styled'

const Section = styled.section`
  margin: 3em auto;

  &:first-of-type {
    margin-top: 0;
  }
`

export default Section
