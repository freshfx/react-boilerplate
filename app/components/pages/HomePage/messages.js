/*
 * HomePage Messages
 *
 * This contains all the text for the HomePage component.
 */
import {defineMessages} from 'react-intl'

export default defineMessages({
  startProjectHeader: {
    defaultMessage: 'Start your next react project in seconds',
    id: 'boilerplate.containers.HomePage.start_project.header'
  },
  startProjectMessage: {
    defaultMessage:
      'A highly scalable, offline-first foundation with the best DX and a focus on performance and best practices',
    id: 'boilerplate.containers.HomePage.start_project.message'
  },
  trymeHeader: {
    defaultMessage: 'Try me!',
    id: 'boilerplate.containers.HomePage.tryme.header'
  }
})
