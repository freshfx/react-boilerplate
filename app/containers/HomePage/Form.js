import styled from '@emotion/styled'

const Form = styled.form`
  margin-bottom: 1em;
`

export default Form
