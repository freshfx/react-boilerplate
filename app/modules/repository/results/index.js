import reducer, {actions, selectors} from './slice'
import saga from './saga'

export {actions, saga, selectors}

export default reducer
