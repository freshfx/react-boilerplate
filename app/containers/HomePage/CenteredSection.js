import styled from '@emotion/styled'

import Section from './Section'

const CenteredSection = styled(Section)`
  text-align: center;
`

export default CenteredSection
