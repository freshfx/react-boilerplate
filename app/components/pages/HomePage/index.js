/*
 * HomePage
 *
 * This is the first thing users see of our App, at the '/' route
 */

import React from 'react'
import {Helmet} from 'react-helmet-async'
import {FormattedMessage} from 'react-intl'
import {useSelector} from 'react-redux'

import H2 from 'components/atoms/H2'
import ReposList from 'components/organisms/ReposList'
import UserNameForm from 'components/organisms/UserNameForm'
import GenericTemplate from 'components/templates/GenericTemplate'
import RepositoryResultsInjector from 'hooks/repository/results/Injector'
import UsernameFormInjector from 'hooks/ui/username-form/Injector'
import useLatestDispatcher from 'hooks/useLatestDispatcher'
import useMount from 'hooks/useMount'
import {actions} from 'modules/repository/results'
import {selectors as usernameFormSelectors} from 'modules/ui/username-form'

import CenteredSection from './CenteredSection'
import Section from './Section'
import messages from './messages'

const HomePage = () => {
  const username = useSelector(usernameFormSelectors.selectUsername)
  const fetchRepositories = useLatestDispatcher({
    action: actions.fetchRepositories
  })

  useMount(() => {
    if (username?.length) {
      fetchRepositories()
    }
  })

  return (
    <GenericTemplate>
      <UsernameFormInjector />
      <RepositoryResultsInjector />
      <Helmet>
        <title>Home Page</title>
        <meta
          name="description"
          content="A React.js Boilerplate application homepage"
        />
      </Helmet>
      <div>
        <CenteredSection>
          <H2>
            <FormattedMessage {...messages.startProjectHeader} />
          </H2>
          <p>
            <FormattedMessage {...messages.startProjectMessage} />
          </p>
        </CenteredSection>
        <Section>
          <H2>
            <FormattedMessage {...messages.trymeHeader} />
          </H2>
          <UserNameForm onSubmit={fetchRepositories} />
          <ReposList />
        </Section>
      </div>
    </GenericTemplate>
  )
}

export default React.memo(HomePage)
