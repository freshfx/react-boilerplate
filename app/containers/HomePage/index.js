/*
 * HomePage
 *
 * This is the first thing users see of our App, at the '/' route
 */

import React from 'react'
import PropTypes from 'prop-types'
import {Helmet} from 'react-helmet-async'
import {FormattedMessage} from 'react-intl'
import {connect} from 'react-redux'
import {compose} from 'redux'
import {createStructuredSelector} from 'reselect'
import noop from 'lodash/noop'
import {
  injectReducer,
  injectSaga
} from 'redux-injectors'

import H2 from 'components/H2'
import ReposList from 'components/ReposList'
import repositoriesReducer, {
  actions as repositoriesActions,
  saga as repositoriesSaga,
  selectors as repositoriesSelectors
} from 'modules/repository/results'
import reducer, {
  actions,
  selectors
} from 'modules/pages/home'

import AtPrefix from './AtPrefix'
import CenteredSection from './CenteredSection'
import Form from './Form'
import Input from './Input'
import Section from './Section'
import messages from './messages'

class HomePage extends React.PureComponent {
  /*
   * When initial state username is not null, submit the form to load repos
   */
  componentDidMount() {
    if (this.props.username && this.props.username.trim().length) {
      this.props.onSubmitForm()
    }
  }

  render() {
    return (
      <article>
        <Helmet>
          <title>Home Page</title>
          <meta
            name="description"
            content="A React.js Boilerplate application homepage"
          />
        </Helmet>
        <div>
          <CenteredSection>
            <H2>
              <FormattedMessage {...messages.startProjectHeader} />
            </H2>
            <p>
              <FormattedMessage {...messages.startProjectMessage} />
            </p>
          </CenteredSection>
          <Section>
            <H2>
              <FormattedMessage {...messages.trymeHeader} />
            </H2>
            <Form onSubmit={this.props.onSubmitForm}>
              <label htmlFor="username">
                <FormattedMessage {...messages.trymeMessage} />
                <AtPrefix>
                  <FormattedMessage {...messages.trymeAtPrefix} />
                </AtPrefix>
                <Input
                  id="username"
                  type="text"
                  placeholder="mxstbr"
                  value={this.props.username}
                  onChange={this.props.onChangeUsername}
                />
              </label>
            </Form>
            <ReposList
              error={this.props.error}
              loading={this.props.isLoading}
              repos={this.props.repositories}
            />
          </Section>
        </div>
      </article>
    )
  }
}

HomePage.propTypes = {
  error: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.bool
  ]),
  isLoading: PropTypes.bool,
  onChangeUsername: PropTypes.func,
  onSubmitForm: PropTypes.func,
  repositories: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.bool
  ]),
  username: PropTypes.string
}

HomePage.defaultProps = {
  error: false,
  isLoading: false,
  onChangeUsername: noop,
  onSubmitForm: noop,
  repositories: false,
  username: ''
}

const mapStateToProps = createStructuredSelector({
  error: repositoriesSelectors.selectError,
  isLoading: repositoriesSelectors.selectIsLoading,
  repositories: repositoriesSelectors.selectRepositories,
  username: selectors.selectUsername
})

const mapDispatchToProps = dispatch => ({
  onChangeUsername: event =>
    dispatch(actions.changeUsername({username: event.target.value})),
  onSubmitForm: event => {
    if (typeof event !== 'undefined' && event.preventDefault) {
      event.preventDefault()
    }
    dispatch(repositoriesActions.loadRepositories())
  }
})

const withConnect = connect(mapStateToProps, mapDispatchToProps)
const withReducer = injectReducer(reducer)
const withRepositoriesReducer = injectReducer(repositoriesReducer)
const withRepositoriesSaga = injectSaga(repositoriesSaga)

export {
  HomePage,
  mapDispatchToProps
}

export default compose(
  withConnect,
  withReducer,
  withRepositoriesReducer,
  withRepositoriesSaga
)(HomePage)
