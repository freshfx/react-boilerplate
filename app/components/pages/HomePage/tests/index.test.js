/**
 * Test the HomePage
 */

import React from 'react'
import {fireEvent} from '@testing-library/react'

import configureStore from 'configure-store'
import HomePageInjector from 'hooks/ui/username-form/Injector'
import {actions} from 'modules/ui/username-form'
import {actions as repositoriesActions} from 'modules/repository/results'
import render from 'utils/test-utils/custom-render'
import renderInjectors from 'utils/test-utils/render-injectors'
import history from 'utils/history'

import HomePage from '../index'

jest.mock('components/organisms/UserNameForm', () => ({onSubmit}) => (
  <form data-testid="username-form" onSubmit={() => onSubmit()} />
))

const store = configureStore({}, history)
jest.spyOn(store, 'dispatch')

const options = {wrapperProps: {store}}
const renderComponent = () => render(<HomePage />, options)

describe('HomePage', () => {
  beforeAll(() => {
    renderInjectors(<HomePageInjector />, options)
  })

  beforeEach(() => {
    store.dispatch.mockClear()
  })

  it('should match the snapshot', () => {
    const {container} = renderComponent()
    expect(container).toMatchSnapshot()
  })

  describe('events', () => {
    describe('mount', () => {
      it('should dispatch a loadRepositories action if there is a username available', () => {
        store.dispatch(actions.changeUsername({username: 'mxstbr'}))
        store.dispatch.mockClear()
        renderComponent()
        expect(store.dispatch).toHaveBeenCalledTimes(1)
        expect(store.dispatch).toHaveBeenCalledWith(
          repositoriesActions.loadRepositories()
        )
        store.dispatch(actions.changeUsername({username: ''}))
      })

      it('should not dispatch a loadRepositories action if there is no username', () => {
        renderComponent()
        expect(store.dispatch).not.toHaveBeenCalled()
      })
    })

    describe('submit form', () => {
      it('should dispatch a loadRepositories action', () => {
        const {getByTestId} = renderComponent()
        fireEvent.submit(getByTestId('username-form'))
        expect(store.dispatch).toHaveBeenCalledTimes(1)
        expect(store.dispatch).toHaveBeenCalledWith(
          repositoriesActions.loadRepositories()
        )
      })
    })
  })
})
