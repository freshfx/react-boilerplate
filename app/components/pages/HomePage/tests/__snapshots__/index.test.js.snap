// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage should match the snapshot 1`] = `
.emotion-0 {
  margin: 3em auto;
  text-align: center;
}

.emotion-0:first-of-type {
  margin-top: 0;
}

.emotion-2 {
  font-size: 1.5em;
}

.emotion-4 {
  margin: 3em auto;
}

.emotion-4:first-of-type {
  margin-top: 0;
}

<div>
  <article>
    <div>
      <section
        class="emotion-0 emotion-1"
      >
        <h2
          class="emotion-2 emotion-3"
        >
          Start your next react project in seconds
        </h2>
        <p>
          A highly scalable, offline-first foundation with the best DX and a focus on performance and best practices
        </p>
      </section>
      <section
        class="emotion-4 emotion-5"
      >
        <h2
          class="emotion-2 emotion-3"
        >
          Try me!
        </h2>
        <form
          data-testid="username-form"
        />
      </section>
    </div>
  </article>
</div>
`;
